gravityflow/gravityflowincomingwebhook
# Response Mapping

**Automated Test Instructions**
* Checkout branch
* Prepare your environment for codeception execution
* run test cases and test happy

**Manual Test Instructions**

* Checkout branch and ensure you have Gravity Forms Rest API V1 or V2
* Import the form(s) in /tests/acceptance-tests/_data/forms/
* Preview the form and submit a single entry
* Go to Workflow Inbox for administrator - approve entry and take note of entry ID
* Use an API test tool to make a request to your endpoint. V2 example: http://local.test/wp-json/gf/v2/entries/[entry ID]/workflow-hooks (updating entry ID). Refer to tests/acceptance-tests/acceptance for the associated cept test that contains related $data and sendPOST vars to emulate. 

Example:

```
{
	"workflow-api-key":"e0631fbf04c45d163da0fc0b4b271655",
	"workflow-api-secret":"2ee9ed974258b984bdbe3bef0f14de2f",
	"1-text":"raw-json",
	"2-radio":"The Mall",
	"3-paragraph":"Large text field mapped via raw json",
	"4-checkboxes": [{"id":"4.1","value":"Alberta"},{"id":"4.2","value":"British Columbia"},{"id":"4.9","value":"Ontario"},{"id":"4.4","value":""},{"id":"4.5","value":""}],
	"43-cb-single": "",
	"6-multiselect": ["Alberta","British Columbia","Ontario"],
	"5-dropdown":"Second Choice",
	"7-number": "42",
	"8-hidden": "Incoming API Hidden Value",
	"11-name-full": [{"id":"11.3","value":"Mickey"},{"id":"11.6","value":"Mantle"}],
	"11-name-middle": "Baseball",
	"12-date": "01/31/2018",
	"13-time": "9:50 AM",
	"14-phone": "987-654-3210",
	"15-address-full":[{"id":"15.1","value":"Apt 1"},{"id":"15.2","value":"2nd Street"}],
	"15-address-city": "New York",
	"16-website": "https://gravityflow.io",
	"17-email": "support@gravityflow.io",
	"32-assignee": "user_id|2",
	"33-discussion": "A new detail from the incoming API call",
	"34-role": "author",
	"35-user": "2"
}

```

**Areas for particular review / discussion in PR**
- Action / Filter consideration
Example: Filter in add_mapping_to_entry for field types that are skipped. Doing so should also have the step settings match through get_field_map_choices exclude_field_types array.

- Security consideration
All fields pass should be passing through their respective sanitize_entry_value either directly or by other calls made before updating entry in add_mapping_to_entry function. Other sanitization/validation review appreciated.

**To Do:**
* Update Documentation and examples of how to use and test the mapping field functionality
